### ISSUE TITLE
Admin Account Script Bugs

### ISSUE THREAD

**Issue by chancecardona** [SixK/coco-annotator-ng#6]
Hi again, thanks so much for the responsiveness so far. 
Currently am running into an error with creating any user accounts:
```
annotator_webclient  | [2024-05-21 22:49:20 +0000] [9] [ERROR] Exception on /api/admin/user/ [POST]
annotator_webclient  | Traceback (most recent call last):
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 880, in full_dispatch_request
annotator_webclient  |     rv = self.dispatch_request()
annotator_webclient  |          ^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 865, in dispatch_request
annotator_webclient  |     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_restx/api.py", line 402, in wrapper
annotator_webclient  |     resp = resource(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/views.py", line 110, in view
annotator_webclient  |     return current_app.ensure_sync(self.dispatch_request)(**kwargs)  # type: ignore[no-any-return]
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_restx/resource.py", line 41, in dispatch_request
annotator_webclient  |     resp = meth(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_login/utils.py", line 290, in decorated_view
annotator_webclient  |     return current_app.ensure_sync(func)(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/workspace/webserver/api/admin.py", line 75, in post
annotator_webclient  |     user.password = generate_password_hash(args.get('password'), method='sha256')
annotator_webclient  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/werkzeug/security.py", line 101, in generate_password_hash
annotator_webclient  |     h, actual_method = _hash_internal(method, salt, password)
annotator_webclient  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/werkzeug/security.py", line 66, in _hash_internal
annotator_webclient  |     raise ValueError(f"Invalid hash method '{method}'.")
annotator_webclient  | ValueError: Invalid hash method 'sha256'.
annotator_webclient  | [2024-05-21 22:49:30 +0000] [9] [DEBUG] POST /api/admin/user/
annotator_webclient  | [2024-05-21 22:49:30 +0000] [9] [ERROR] Exception on /api/admin/user/ [POST]
annotator_webclient  | Traceback (most recent call last):
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 880, in full_dispatch_request
annotator_webclient  |     rv = self.dispatch_request()
annotator_webclient  |          ^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 865, in dispatch_request
annotator_webclient  |     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_restx/api.py", line 402, in wrapper
annotator_webclient  |     resp = resource(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/views.py", line 110, in view
annotator_webclient  |     return current_app.ensure_sync(self.dispatch_request)(**kwargs)  # type: ignore[no-any-return]
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_restx/resource.py", line 41, in dispatch_request
annotator_webclient  |     resp = meth(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_login/utils.py", line 290, in decorated_view
annotator_webclient  |     return current_app.ensure_sync(func)(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/workspace/webserver/api/admin.py", line 75, in post
annotator_webclient  |     user.password = generate_password_hash(args.get('password'), method='sha256')
annotator_webclient  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/werkzeug/security.py", line 101, in generate_password_hash
annotator_webclient  |     h, actual_method = _hash_internal(method, salt, password)
annotator_webclient  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/werkzeug/security.py", line 66, in _hash_internal
annotator_webclient  |     raise ValueError(f"Invalid hash method '{method}'.")
annotator_webclient  | ValueError: Invalid hash method 'sha256'.
annotator_webclient  | [2024-05-21 22:49:33 +0000] [9] [DEBUG] POST /api/admin/user/
annotator_webclient  | [2024-05-21 22:49:33 +0000] [9] [ERROR] Exception on /api/admin/user/ [POST]
annotator_webclient  | Traceback (most recent call last):
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 880, in full_dispatch_request
annotator_webclient  |     rv = self.dispatch_request()
annotator_webclient  |          ^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 865, in dispatch_request
annotator_webclient  |     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_restx/api.py", line 402, in wrapper
annotator_webclient  |     resp = resource(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask/views.py", line 110, in view
annotator_webclient  |     return current_app.ensure_sync(self.dispatch_request)(**kwargs)  # type: ignore[no-any-return]
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_restx/resource.py", line 41, in dispatch_request
annotator_webclient  |     resp = meth(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/flask_login/utils.py", line 290, in decorated_view
annotator_webclient  |     return current_app.ensure_sync(func)(*args, **kwargs)
annotator_webclient  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/workspace/webserver/api/admin.py", line 75, in post
annotator_webclient  |     user.password = generate_password_hash(args.get('password'), method='sha256')
annotator_webclient  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/werkzeug/security.py", line 101, in generate_password_hash
annotator_webclient  |     h, actual_method = _hash_internal(method, salt, password)
annotator_webclient  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
annotator_webclient  |   File "/usr/local/lib/python3.11/site-packages/werkzeug/security.py", line 66, in _hash_internal
annotator_webclient  |     raise ValueError(f"Invalid hash method '{method}'.")
annotator_webclient  | ValueError: Invalid hash method 'sha256'.
```

When I try to do the workaround listed in the readme I get the following error:
`$python change_password_hash_type.py`
```
{'name': 'Remove Current Polygon', 'keys': ['escape']}, {'name': 'Increase Radius', 'keys': ['[']}, {'name': 'Decrease Radius', 'keys': [']']}, {'name': 'Increase Radius', 'keys': ['[']}, {'name': 'Decrease Radius', 'keys': [']']}, {'name': 'Subtract Selection', 'keys': ['shift', 'click']}]}}, 'permissions': [], 'last_seen': datetime.datetime(2024, 5, 21, 22, 50, 20, 299000)}
old_hash: None
Traceback (most recent call last):
  File "/home/chance/Repos/coco-annotator-ng/change_password_hash_type.py", line 67, in <module>
    new_hash = verify_password_and_generate_hash(password, old_hash)
  File "/home/chance/Repos/coco-annotator-ng/change_password_hash_type.py", line 23, in verify_password_and_generate_hash
    if check_password_hash(hashed_password, password):
  File "/home/chance/.pyenv/versions/3.10.4/lib/python3.10/site-packages/werkzeug/security.py", line 215, in check_password_hash
    if pwhash.count("$") < 2:
AttributeError: 'NoneType' object has no attribute 'count'
```

**Comment by SixK** [comment]
Problem to create new user account should be fixed by commit 3c034a4e64703563ff285cd2e7a1c864914f3e69
Sorry for this bug, I didn't noticed I was still using sha256 instead of pbkdf2:sha256 in admin.py file.

Thank's for your feedbacks, you help me to stabilize coco-annotator-ng.

change_password_hash_type.py script is only help to migrate users accounts for a database created using JsBroks official coco-annotator (using werkzeug 2) to coco-annotator-ng (using werkzeug 3). 
Werkzeug 3 introduced a breaking change on password encryption method hash. It now use pbkdf2:sha256 method instead of sha256. 
With this script you only can change user password if you know user password and you have to modify the script to change username and password.

**Comment by chancecardona** [comment]
Great thank you so much. Tested and this works on my end. :clinking_glasses: :fireworks:
